rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/CompactView.java

```diff
@@ -166,269 +165,107 @@ private void changeSelection(String name) {
+            if (nodeSource.isRemoved()) {
+                if (flow.isNodeSourceDrawn(nodeSource.getSourceName())) {
+                    flow.remove(nodeSource);
+                }
+            } else {
+                if (!flow.isNodeSourceDrawn(nodeSource.getSourceName())) {
+                    Tile nsTile = new Tile(nodeSource);
+                    flow.drawNodeSource(nsTile);
                 }
 
-                /* hosts */
-                for (Host h : ns.getHosts().values()) {
-                    String hostName = h.getHostName();
-
-                    Host oldHost = (oldNs != null) ? oldNs.getHosts().get(hostName) : null;
-                    /* new host */
-                    if (oldHost == null) {
-                        NodeTile hostTile = new NodeTile(h);
-                        int i = curTiles.indexOf(nsName) + 1 + ns.getDeploying().size();
-                        flow.insert(hostTile, i);
-                        this.curTiles.add(i, h.getId());
-                    }
-                    /* nodes */
-                    for (Node n : h.getNodes().values()) {
-                        String nodeUrl = n.getNodeUrl();
-
-                        Node oldNode = (oldHost != null) ? oldHost.getNodes().get(nodeUrl) : null;
-                        /* new node */
-                        if (oldNode == null) {
-                            NodeTile nodeTile = new NodeTile(n);
-                            int i = curTiles.indexOf(h.getId()) + 1;
-                            flow.insert(nodeTile, i);
-                            this.curTiles.add(i, nodeUrl);
-                        }
-                        /* update old node status */
-                        else {
-                            if (!oldNode.getNodeState().equals(n.getNodeState()) ||
-                                oldNode.isLocked() != n.isLocked()) {
-                                int i = this.curTiles.indexOf(nodeUrl);
-                                NodeTile nt = ((NodeTile) this.flow.getWidget(i));
-                                nt.refresh(n);
-                            }
-                        }
-                    }
+                if (nodeSource.isChanged()) {
+                    flow.redrawNodeSource(nodeSource);
                 }
             }
+        }
+    }
 
-            /*
-             * now remove the difference between the nodes from this method call,
-             * and the ones from the previous call
-             */
-            for (Entry<String, NodeSource> oldNs : this.oldNodes.entrySet()) {
-                /* Keep NS */
-                NodeSource newNs = nodes.get(oldNs.getKey());
-                if (newNs != null) {
-                    for (Entry<String, Node> oldDepl : oldNs.getValue().getDeploying().entrySet()) {
-                        /* Keep deploying Node */
-                        if (newNs.getDeploying().containsKey(oldDepl.getKey())) {
+    private void initializePanel() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 143 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)